src/autorest.bicep/src/schema-generator.ts

```diff
@@ -0,0 +1,729 @@
+        oneOf: oneOf,
+      };
+
+      return {
+        descriptor: {
+          ...descriptor,
+          constantName: undefined,
+        },
+        schema: addResourceTypeAndApiVersion(descriptor, schema, isChildDefinition),
+      };
+    } else {
+      const definition = definitions[0];
+      const schema = processResourceBody(fullyQualifiedType, definition, isChildDefinition);
+      if (!schema) {
+        return null;
+      }
+
+      return {
+        descriptor,
+        schema: addResourceTypeAndApiVersion(descriptor, schema, isChildDefinition),
+      };
+    }
+  }
+
+  function getObjectName(putSchema: ObjectSchema | undefined) {
+    return putSchema ? getSerializedName(putSchema) : undefined;
+  }
+
+  function parseObjectType(putSchema: ObjectSchema | undefined, ancestorsToExclude?: Set<ComplexSchema>) {
+    const combinedSchema = throwIfNull(putSchema);
+    const definitionName = getObjectName(putSchema);
+
+    if (!ancestorsToExclude && schemaData.definitions[definitionName]) {
+      // if we're building a discriminated subtype, we're going to be missing the base properties
+      // so construct the type on-the-fly, and don't cache it globally
+      return {
+        '$ref': `#/definitions/${definitionName}`,
+      };
+    }
+
+    const lookupParentDictionary = (s: ObjectSchema | undefined) => chain(s?.parents?.all || [])
+      .filter(s => {
+        if (ancestorsToExclude && ancestorsToExclude.has(s)) {
+          return false;
+        }
+
+        return s instanceof DictionarySchema;
+      })
+      .map(s => s as DictionarySchema)
+      .first()
+      .value();
+
+    const putParentDictionary = lookupParentDictionary(putSchema);
+    const additionalProperties = putParentDictionary ? parseType(putParentDictionary?.elementType, true) : undefined;
+
+    const definition = createObject({}, additionalProperties);
+    definition.description = getTypeDescription(combinedSchema);
+    if (!ancestorsToExclude) {
+      // cache the definition so that it can be re-used
+      schemaData.definitions[definitionName] = definition;
+    }
+
+    // Only make a distinction between what's defined on PUT vs GET if we're dealing with a synthetic object or a discriminated subtype.
+    // If the schema on both PUT and GET is the same named object (or if one of the two is undefined),
+    // use the combined schema as both GET and PUT schemata to prevent ReadOnly/WriteOnly flags from trickling down
+    // to object properties (which is problematic if shapes are reused across resources)
+    //
+    // For discriminated subtypes, Bicep's type system does not have a great way to communicate which variants are available on read vs write, but this
+    // can be communicated on variant properties. NB: `putSchema` and `getSchema` will only be different in a discriminated subtype if the discriminated
+    // object was synthetic.
+    const schemaForPut = ancestorsToExclude ? putSchema : combinedSchema;
```

Comment:
`putSchema` and `combinedSchema` should be the same, no? Since schema generation is only concern with the PUT, I don't think you need to account for synthetic object types.

Reply:
Good point - updated.